5747 PID Segment from FHIR to HL7

TurkeyFried opened the pull request:
This PR generates the HL7 PID segment from FHIR data.

Test Steps:
1. Check out the branch for this PR.
2. Run the following command to convert a FHIR bundle to HL7.  
`./prime fhirdata --input-file "src/testIntegration/resources/datatests/HL7_to_FHIR/sample_co_1_20220518-0001.fhir" -s metadata/hl7_mapping/ORU_R01/ORU_R01-base.yml`
3. Verify an HL7 message is generated from the conversion and includes the PID segments.
4. Look at the source HL7 used at `src/testIntegration/resources/datatests/HL7_to_FHIR/sample_co_1_20220518-0001.hl7` and compare the data for the PID segments.
5. Verify both PID segments are identical, except for these caveats:
 ```
The following fields are not supported on HL7 2.5.1 and thus are ignored on the conversion: PID-10.7, PID-10.8, PID-22.7, PID-22.8, PID-33.7, PID-33.8
PID-17.1 has a known issue when converting to FHIR (see #6483) and thus will be skipped on the output.
```

## Changes
- Added config to handle PID conversion from the FHIR data.
- Made a few updates on the HL7 to FHIR conversion to properly store some data.

## Checklist

### Testing
- [x] Tested locally?
- [ ] Ran `./prime test` or `./gradlew testSmoke` against local Docker ReportStream container?
- [ ] (For Changes to /frontend-react/...) Ran `npm run lint:write`? 
- [ ] Added tests?

### Process
- [ ] Are there licensing issues with any new dependencies introduced?
- [ ] Includes a summary of what a code reviewer should test/verify?
- [ ] Updated the release notes?
- [ ] Database changes are submitted as a separate PR?
- [ ] DevOps team has been notified if PR requires ops support?

## Linked Issues
- #5747 

## To Be Done
*Create GitHub issues to track the work remaining, if any*

## Specific Security-related subjects a reviewer should pay specific attention to
- Does this PR introduce new endpoints?
    - new endpoint A
    - new endpoint B
- Does this PR include changes in authentication and/or authorization of existing endpoints?
- Does this change introduce new dependencies that need vetting?
- Does this change require changes to our infrastructure?
- Does logging contain sensitive data?
- Does this PR include or remove any sensitive information itself?

If you answered '_yes_' to any of the questions above, conduct a detailed Review that addresses at least:

- What are the potential security threats and mitigations? Please list the _STRIDE_ threats and how they are mitigated
    - **S**poofing (faking authenticity)
        - Threat _T_, which could be achieved by _A_, is mitigated by _M_
    - **T**ampering (influence or sabotage the integrity of information, data, or system)
    - **R**epudiation (the ability to dispute the origin or originator of an action)
    - **I**nformation disclosure (data made available to entities who should not have it)
    - **D**enial of service (make a resource unavailable)
    - **E**levation of Privilege (reduce restrictions that apply or gain privileges one should not have)
- Have you ensured logging does not contain sensitive data?
- Have you received any additional approvals needed for this change?

carlosfelix2 commented on prime-router/metadata/fhir_mapping/hl7/datatype/Identifier_SystemID.yml:
```diff
@@ -104,4 +104,25 @@ extension_4:
+  valueOf: extension/Extension
+  expressionType: resource
+  vars:
+    value: String, PID.3.4.1
```
Do not specify direct fields in data types.  This must be done at a higher level.  I am pretty sure this needs to be in the value field

TurkeyFried replied:
Not sure what to change here. Let's talk it on monday

carlosfelix2 commented on prime-router/metadata/hl7_mapping/ORU_R01/base/telecom.yml:
```diff
@@ -23,4 +23,8 @@ elements:
\ No newline at end of file
+    hl7Spec: [ '%{hl7TelecomField}-7' ]
+
+  - name: any-text
```
rename to text

TurkeyFried replied:
fixed

carlosfelix2 commented on prime-router/metadata/hl7_mapping/ORU_R01/base/xad-extended-address.yml:
```diff
@@ -1,36 +1,79 @@
+  - name: street-address-ordered
+    condition: '%{orderIndex} NOT_NULL && %resource.line.count() > 0'
+    value: [ '%resource.line[0]' ]
+    hl7Spec: [ '%{fieldPath}-1' ]
```
This is confusing having fieldPath and hl7AddressField

TurkeyFried replied:
fixec

carlosfelix2 commented on prime-router/metadata/hl7_mapping/ORU_R01/base/xad-extended-address.yml:
```diff
@@ -1,36 +1,79 @@
     value: [ '%resource.line[0]' ]
     hl7Spec: [ '%{hl7AddressField}-1' ]
 
+  - name: street-address-ordered
```
There seem to be duplicate fields here. Line 5 and 11.  Some more below

TurkeyFried replied:
fixed

carlosfelix2 commented on prime-router/src/main/kotlin/fhirengine/translation/hl7/utils/ConstantResolver.kt:
```diff
@@ -200,6 +203,32 @@ object CustomFHIRFunctions {
+    fun getYesNoValue(focus: MutableList<Base>): MutableList<Base> {
+        return when (focus[0].toString()) {
+            "true" -> mutableListOf(StringType("Y"))
+            else -> mutableListOf(StringType("N"))
```
else needs to be empty string.  You do not want to add data to a field that does not have a value

TurkeyFried replied:
Fixed

TurkeyFried commented on prime-router/metadata/hl7_mapping/ORU_R01/base/cx-extended-composite-id.yml:
```diff
@@ -0,0 +1,22 @@
+name: ORU-R01-cx-extended-composite-id
+constants:
+  fieldPath: /PATIENT_RESULT/PATIENT/(%{hl7CXField})(%{idIndex})
```
fixed

TurkeyFried commented on prime-router/metadata/hl7_mapping/ORU_R01/base/hd-hierarchic-designator-facility.yml:
```diff
@@ -0,0 +1,16 @@
+    hl7Spec: [ '%{hl7HDField}-2' ]
+
+  - name: authority-universal-id-type
+    condition: '%resource.extension(%`rsext-assigning-facility-namespace-id`).exists()'
```
fixed

TurkeyFried commented on prime-router/metadata/hl7_mapping/ORU_R01/base/patient.yml:
```diff
@@ -1,12 +1,166 @@
-    schema: contact
\ No newline at end of file
+    constants:
+      hl7NameField: '/PATIENT_RESULT/PATIENT/PID-5'
```
fixed

TurkeyFried commented on prime-router/metadata/hl7_mapping/ORU_R01/base/hd-hierarchic-designator-authority.yml:
```diff
@@ -0,0 +1,16 @@
+name: ORU-R01-hd-hierarchic-designator-authority
+elements:
+  - name: authority-namespace-id
+    condition: '%resource.extension(%`rsext-assigning-authority-namespace-id`).exists()'
```
fixed

carlosfelix2 commented on prime-router/metadata/fhir_mapping/hl7/datatype/Identifier_SystemID.yml:
```diff
@@ -104,4 +104,15 @@ extension_4:
\ No newline at end of file
+    urlValue: assigning-facility-namespace-id
+
+# Assigning authority namespace
```
Remove these commented out lines

carlosfelix2 commented on prime-router/src/testIntegration/resources/datatests/HL7_to_FHIR/sample_co_1_20220518-0001.fhir:
```diff
@@ -410,7 +410,7 @@
         },
         "system" : "urn:id:RML",
-        "value" : "1033122247"
+        "value" : "RML"
```
This is a problem

carlosfelix2 commented on prime-router/src/testIntegration/resources/datatests/HL7_to_FHIR/sample_co_1_20220518-0001.fhir:
```diff
@@ -697,7 +697,7 @@
         },
         "system" : "urn:id:RML",
-        "value" : "1033122247"
+        "value" : "RML"
```
And here

carlosfelix2 commented on prime-router/src/testIntegration/resources/datatests/HL7_to_FHIR/sample_co_1_20220518-0001.fhir:
```diff
@@ -1016,7 +1016,7 @@
         },
         "system" : "urn:id:RML",
-        "value" : "1033122247"
+        "value" : "RML"
```
And here

carlosfelix2 commented on prime-router/src/testIntegration/resources/datatests/HL7_to_FHIR/sample_co_1_20220518-0001.fhir:
```diff
@@ -1278,7 +1278,7 @@
         },
         "system" : "urn:id:RML",
-        "value" : "1740336429"
+        "value" : "RML"
```
and here

carlosfelix2 commented on prime-router/src/testIntegration/resources/datatests/HL7_to_FHIR/sample_co_1_20220518-0001.fhir:
```diff
@@ -1545,7 +1545,7 @@
         },
         "system" : "urn:id:RML",
-        "value" : "1740336429"
+        "value" : "RML"
```
here

carlosfelix2 commented on prime-router/src/testIntegration/resources/datatests/HL7_to_FHIR/sample_co_1_20220518-0001.fhir:
```diff
@@ -1818,7 +1818,7 @@
         },
         "system" : "urn:id:RML",
-        "value" : "1740336429"
+        "value" : "RML"
```
here

carlosfelix2 commented on prime-router/src/testIntegration/resources/datatests/HL7_to_FHIR/sample_co_1_20220518-0001.fhir:
```diff
@@ -2085,7 +2085,7 @@
         },
         "system" : "urn:id:RML",
-        "value" : "1740336429"
+        "value" : "RML"
```
here

carlosfelix2 commented on prime-router/src/testIntegration/resources/datatests/HL7_to_FHIR/sample_co_1_20220518-0001.fhir:
```diff
@@ -9330,7 +9330,7 @@
         },
         "system" : "http://hl7.org/fhir/sid/us-npi",
-        "value" : "1871759365"
+        "value" : "NPI"
```
This is a problem

carlosfelix2 commented on prime-router/src/testIntegration/resources/datatests/HL7_to_FHIR/sample_co_1_20220518-0001.fhir:
```diff
@@ -11202,7 +11202,7 @@
         },
         "system" : "urn:id:Medical_Center_of_Au",
-        "value" : "1437205069"
+        "value" : "Medical Center of Au"
```
Looks to be a problem too

TurkeyFried commented:
Fixed and resolved the mapping issues. There was a change on Identifier that caused this. Reverting it kept the new mapping working and fixed everything.

PatriciaPerozoUSDS commented on prime-router/src/main/kotlin/fhirengine/translation/hl7/utils/ConstantResolver.kt:
```diff
@@ -34,7 +34,10 @@ data class CustomContext(val bundle: Bundle, val constants: MutableMap<String, S
-                val newContext = previousContext.copy()
+                val newContext = CustomContext(
+                    previousContext.bundle,
+                    previousContext.constants.toMap().toMutableMap() // This makes a copy of the map
```
Thank you for adding the comment here that was helpful

PatriciaPerozoUSDS commented on prime-router/metadata/hl7_mapping/ORU_R01/base/message-header.yml:
```diff
@@ -19,8 +19,8 @@ elements:
   - name: sending-facility-universal-id
-    condition: '%resource.sender.resolve().identifier.value is Oid'
-    value: [ '%resource.sender.resolve().identifier.value.toString().replace(''urn:oid:'','''')' ]
+    condition: '%resource.sender.resolve().identifier[0].value.startsWith(''urn:oid:'')'
```
Is it possible to pull this without counting on it being the first element? I will test on my end as well

PatriciaPerozoUSDS commented on prime-router/metadata/hl7_mapping/ORU_R01/base/patient.yml:
```diff
@@ -1,12 +1,177 @@
+    hl7Spec: [ '%{hl7PIDField}-7' ]
+
+  - name: patient-sex
+    value: [ '%resource.gender.GetAdministrativeGenderCode()' ]
```
I think its worth checking existence before calling a custom function on it
